src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1285,21 +1285,112 @@ internal bool HasCollectionExpressionApplicableAddMethod(SyntaxNode syntax, Type
+
+                // Value parameter must be by value, in, or ref readonly.
+                if (parameters[valueIndex].RefKind is not (RefKind.None or RefKind.In or RefKind.RefReadOnlyParameter))
+                {
+                    return false;
+                }
+
+                // If the method is generic, can the type arguments be inferred from the one or two arguments?
+                var typeParameters = method.TypeParameters;
+                if (typeParameters.Length > 0)
+                {
+                    var usedParameterTypes = parameters.Slice(0, requiredLength).SelectAsArray(p => p.Type);
+                    foreach (var typeParameter in typeParameters)
+                    {
+                        if (!usedParameterTypes.Any((parameterType, typeParameter) => parameterType.ContainsTypeParameter(typeParameter), typeParameter))
+                        {
+                            // The type parameter does not appear in any of the parameter types.
+                            return false;
+                        }
+                    }
+                }
+
+                return true;
+            }
+        }
+
         /// <summary>
         /// If the element is from a collection type where elements are added with collection initializers,
         /// return the argument to the collection initializer Add method or null if the element is not a
         /// collection initializer node. Otherwise, return the element as is.
         /// </summary>
-        internal static BoundExpression? GetUnderlyingCollectionExpressionElement(BoundCollectionExpression expr, BoundExpression? element)
+        internal static BoundExpression GetUnderlyingCollectionExpressionElement(BoundCollectionExpression expr, BoundExpression element, bool throwOnErrors)
         {
             if (expr.CollectionTypeKind is CollectionExpressionTypeKind.ImplementsIEnumerable)
             {
-                return element switch
+                switch (element)
                 {
-                    BoundCollectionElementInitializer collectionInitializer => getCollectionInitializerElement(collectionInitializer),
-                    BoundDynamicCollectionElementInitializer dynamicInitializer => dynamicInitializer.Arguments[0],
-                    _ => null,
-                };
+                    case BoundCollectionElementInitializer collectionInitializer:
+                        return getCollectionInitializerElement(collectionInitializer);
+                    case BoundDynamicCollectionElementInitializer dynamicInitializer:
+                        return dynamicInitializer.Arguments[0];
+                }
+                if (throwOnErrors)
+                {
+                    throw ExceptionUtilities.UnexpectedValue(element);
+                }
+                switch (element)
+                {
+                    case BoundCall call:
+                        return call.Arguments[call.InvokedAsExtensionMethod ? 1 : 0];
```

Comment:
>return call.Arguments[call.InvokedAsExtensionMethod ? 1 : 0]; [](http://example.com/codeflow?start=24&length=61)

This logic looks suspicious. It is not obvious what are we trying to accomplish and why. It looks like ```BoundCollectionExpression``` holds onto the original ```UnconvertedCollectionExpression``` node. Could we. perhaps, use information from that node instead? Not sure what do we want to do with expressions that do not have a natural type though.

Reply:
It feels like it would be better to place this special logic into ```CreateBoundCollectionExpressionSpreadElement```, where it would make sense as an error recovery logic. It doesn't feel like it belongs in this general helper because we don't really know where the ```element``` node is coming from and whether it is appropriate to dig into it this way.

Reply:
This case is to unwrap the `Add()` argument when [`bindCollectionInitializerElementAddMethod`](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs#L6000) fails in overload resolution with one or more inapplicable or ambiguous `Add` methods. We hit this case for spreads and for regular elements.

Reply:
> We hit this case for spreads and for regular elements.

Thanks for the clarification. Consider adding a comment.